woocommerce/google-listings-and-ads
# Improve WordPress.com account handling

### Changes proposed in this Pull Request:

This PR is intended to improve the handling of the WordPress.com connection status. There are several fixes within this PR that handled various scenarios.

#### Uncaught Fatal Error
When we reset all the plugin data using the snippet `define( 'WC_GLA_REMOVE_ALL_DATA', true );`
And at the same time disconnect the WordPress.com connection externally. We end up with a scenario where it assumes the connection is still intact but it hasn't accepted the TOS conditions yet. This triggers a request to WCS which fails because the token is no longer valid, and in turn triggers a ReconnectAccount exception. In this PR we make sure this exception is caught and handled.

Closes #2285 

#### Disconnect during onboarding
If we trigger an external disconnect during onboarding, we end up with a situation where the local token in the DB is still valid but once an external request is sent it returns a response which triggers a ReconnectAccount exception.
The UI tries to resolve this by checking if the WordPress.com account is still connected. Since this only checks the local token it assumes it's connected.

In this PR we resolve that scenario by doing a more extensive check with the token. We both confirm it's available and send a request to a token endpoint to validate the token hasn't been invalidated externally. This ensures the UI gets a response that the account is no longer connected and can once again show the connect button.

Closes #2284

#### Disconnected after onboarding
If we trigger an external disconnect after onboarding, we end up with the same situation where the local token in the DB is still valid but once an external request is sent it returns a response which triggers a ReconnectAccount exception.

In this PR we again resolve that both by doing an external request to validate the token, and by updating the option `JETPACK_CONNECTED` to reflect the actual status. This allows the UI to catch the error and stay on the reconnect page without assuming it's connected.

Closes #1972

#### Remove disconnect button
This PR also removes the disconnect button on the Connection Test page, as it can only perform a "soft connect", which will always show WooCommerce is connected so we are unable to disconnect.
The intention is to handle this better within WooCommerce or some other higher level, at the moment the Jetpack plugin can be installed to disconnect fully.

Closes #2118

#### Jetpack packages update
To ensure we are working with the latest version of the Jetpack connect I also updated the composer packages. According to the changelogs there are mostly minor fixes. The major version bump is for dropping PHP < 7. Since we have a minimum of PHP 7.4 those changes don't affect anything.
`composer update automattic/jetpack* --with-all-dependencies`
```
- Upgrading automattic/jetpack-a8c-mc-stats (v1.4.22 => v2.0.0)
- Upgrading automattic/jetpack-admin-ui (v0.2.24 => v0.3.2)
- Upgrading automattic/jetpack-autoloader (v2.12.0 => v3.0.2)
- Upgrading automattic/jetpack-config (v1.15.4 => v2.0.0)
- Upgrading automattic/jetpack-connection (v1.59.0 => v2.3.2)
- Upgrading automattic/jetpack-constants (v1.6.23 => v2.0.0)
- Upgrading automattic/jetpack-redirect (v1.7.27 => v2.0.0)
- Upgrading automattic/jetpack-roles (v1.4.25 => v2.0.0)
- Upgrading automattic/jetpack-status (v1.18.5 => v2.1.0)
```

### Detailed test instructions:

#### Uncaught Fatal Error
1. Go through the onboarding steps to connect at least the accounts in step 1
2. Abort onboarding
3. Visit the page /wp-admin/options.php (or directly in the DB) and set the option `gla_wp_tos_accepted` to 0
4. Go to the URL `jptools.wordpress.com/debug/` and find your site
5. Scroll down to the section Blog Details and click Disconnect Jetpack
6. Restart onboarding from the beginning
7. Confirm we are able to connect a WordPress.com account and continue onboarding without issues

#### Disconnect during onboarding
1. Go through the onboarding steps to connect at least the accounts in step 1
2. Abort onboarding
3. Go to the URL `jptools.wordpress.com/debug/` and find your site
4. Scroll down to the section Blog Details and click Disconnect Jetpack
5. Restart onboarding from the beginning
6. Confirm that we see a request to `/wc/gla/jetpack/connected` with a response similar to:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```

#### Disconnect after onboarding
1. Go through all the onboarding steps
2. Go to the URL `jptools.wordpress.com/debug/` and find your site
3. Scroll down to the section Blog Details and click Disconnect Jetpack
4. Refresh one of the GLA pages like Dashboard or Settings
5. Confirm that we are redirected to a reconnect page
![image](https://github.com/woocommerce/google-listings-and-ads/assets/11388669/6c70a5b3-83e4-40dd-95b5-b5c6c05aff3b)
6. Confirm that we see a request to `/wc/gla/jetpack/connected` with a response similar to:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```
7. Try the connect button on the Reconnect Page and confirm that all the pages work again after this is successful.

#### Test Connection Test page
1. Simulate the same scenario as one of the previous where we disconnect the WordPress.com account externally
2. Visit the Connection Test page and expect to see a section with a reconnect button (similar as is done through the UI). Reconnect button should only be visible in cases where we still have a local connection token, but the token is no longer valid.
![image](https://github.com/woocommerce/google-listings-and-ads/assets/11388669/7fac72b8-bcbf-4b24-b505-b23d7c4b5199)
3. Try the status button and expect a response like:
```json
{
    "active": "no",
    "owner": "yes",
    "displayName": "User",
    "email": "user@email.test"
}
```
4. Try the reconnect button and confirm that we can correctly connect again (same as in the UI)

### Changelog entry
* Fix - Improve WordPress.com account handling.
